README.md
# JAK - JAvascriptová Knihovna #

JAK is an open-sourced JavaScript toolkit, created by Seznam.cz, a.s. It aims to simplify development of recurring JS tasks. Several widgets are included. 

This version of JAK is compatible with public released version at [http://jak.seznam.cz/](http://jak.seznam.cz/) (in Czech).

We use JSUnit for testing JAK core library. Test are in /test directory. Testing Request works only with PHP on server side.

JAK is distributed under terms of [MIT licence](http://www.opensource.org/licenses/mit-license.php).


lib/changelog.txt
############################################################
 JAK ChangeLog

  2010-07-26
	kompletní přepsání na verzi 2. Změna jmenného prostoru ze SZN na JAK.
	Informace o migraci je možno nalézt na URL: http://jak.seznam.cz/migration/


  2010-03-09
	decorator.js - SZN.AbstractDecorator (verze: 1.0) SZN.IDecorable (verze: 1.0)
	  - abstraktní třída dekorátoru a rozhranní podporující dekorování.
		Nepodporujeme klasický dekorátor z návrhových vzorů, dekorování provádíme
		pomocí monkey-patch, tedy vlastnosti a metody dekorátoru linkujeme do
		dekorované instance.

  2010-01-26
	browser.js - SZN.Browser (verze: 2.4)
	  - detekce Chrome a FireFox 3.5 a 3.6
	classmaker.js - SZN.ClassMaker (verze: 4.2)
	  - tvorba tříd, rozhraní a namespace je nyní možná (a doporučená) pomocí
		metod makeClass(), makeInterface() a makeStatic(). Navíc přibyla možnost
		vytvořit singleton pomocí metody makeSingleton(), vytvořená třída
		obsahuje vždy metodu getInstance() a nejde zavolat konstruktor.
		V konfiguračním objektu, který je předáván těmto metodám již není
		povinný parametr CLASS, který pozbývá významu. Zpětná kompatibilita
		metody makeClass() je zachována.
	 dom.js - SZN.Dom (verze: 3.6)
	   - Přibyly metody separateCode() a shiftBox(). První umí oddělit JS od
		 HTML, které je jí předané a vrátit je separované. To je vhodné, pokud
		 příchází takovéto HTML jako odpověď na AJAX a my ho připínáme do stránky
		 a chceme kód vykonat. Druhá metoda spočte o kolik je nutné posunout
		 předaný element, aby byl viditelný v okně prohlížeče bez nutnosti
		 scrolovat.
	 events.js - SZN.Events (verze: 2.4)
	   - ostranění kódu specifického pro Operu, chová se standardně.
	 main.js - SZN (verze: 1.1)
	   - definování metody Function.prototype.bind dle návrhu EcmaScript 5 pro
		 prohlížeče, které ji neumí. Funkce SZN.bind tuto nově definovanou
		 funkci využívá.
	 signals.js - SZN.Signals (verze: 1.5)
	   - vnitřní úprava ukládání dat pro rychlejší odebírání posluchačů
	 decorator.js (verze: 1.0)
	   - nová část knihovny.
	   SZN.AbstractDecorator
		 - jedináček (Singleton), z kterého musíme zdědit pokud chceme využít
		   mechanizmus dekorování
	   SZN.AutoDecorator
		 - třída, z které je dobré zdědit iterátory pokud se nechceme starat
		   o proxování volání metod z dekorátoru na dekorovaný objekt. Dědí z
		   SZN.AbstractDecorator
	   SZN.IDecorable
		 - rozhranní, které přidává metodu decorate(), pro jednodušší dekorování
	   Nejedná se o vzor Decorator v pravém slova smyslu, ale o monkeypatch,
	   tedy injektování funkčnosti do původního objektu.

  2009-09-01
	browser.js - SZN.Browser (verze: 2.3)
	  - detekce IE8
	  - detekce verze Opery dle window.opera.version()
	dom.js - SZN.Dom (verze: 3.6)
	  - upravena metoda getBoxPosition(), která nyní počítá i s fixními prvky
	  - přidána metoda writeStyle(), pomocí které je možné přidat styly přímo
		do stránky do elementu STYLE, tedy ne inlinově ke konkrétnímu prvku.
	events.js - SZN.Events (verze: 2.3)
	  - upraven kód pro událost onDomReady tak, aby fungovala i ve stránkách
		tvořených rámy
	  - metoda getInfo() upravena tak, aby vracela setříděné události dle
		prvků, na které jsou navěšeny, tak jak to vyžaduje vizualizátor

  2009-07-23
	object.js
	  - změněno hluboké kopírování objektů

  2009-06-03
	main.js
	  - prototypové metody pro Date nyní formátují názvy dnů a měsíců česky.

  2009-02-23
	classmaker.js - SZN.ClassMaker (verze: 4.1)
	  - zrušena metoda functionToStringTest(), která byla využívána ve volání callSuper().
	  - metoda callSuper() pro volání metody rodíče použává stejné principy jako $super(),
		nicméně její funkčnost zůstala zachována.
	dom.js - SZN.Dom (verze: 3.5)
	  - opravena metoda getBoxPosition() vracející polohu elementu ve stránce.
		V IE ve standardním renderovacím módu se při průchodu nahoru liší odsazení
		body o hodnoty clientLeft a Top, nyní jsou od výsledku odečítány.
	events.js - SZN.Events (verze: 2.2)
	  - přidána metoda getInfo() pro získání informací o navěšených metodách ve
		formátu vhodném pro bookmarklet Visual Event: http://www.sprymedia.co.uk/article/Visual+Event
	main.js
	  - doplnění prototypových metod lpad() a rpad() do prototypového objektu String,
		metody umožňují přidat zleva, či zprava řetězce znaky na požadovanou délku.
		Chování shodné jako v metodě str_pad() v PHP.
	  - doplnění prototypové metody format() do prototypového objektu Date.
		Metoda umožňuje formátovat datum shodným způsobem jako funkce date() v PHP.


  2009-01-28
	classmaker.js - SZN.ClassMaker (verze: 4.0)
	  - zrušena metoda jsonToClass()
	  - vylepšen způsob dědění prototypu pro test instanceof tak, že již není
		nutné si vytvářet vlastnost třídy makeNothing
	  - vylepšena metoda callSuper(), pro volání metody předka. V jistých případech,
		kdy voláme metodu předka, ale předek metodu zdědil, pak se mohlo volání zacyklit.
		Problém vyřešen tagováním kopírovaných prototypových metod.
	  - každá třída vytvořená ClassMakerem nově obsahuje také metodu $super(),
		která by v blízké budoucnosti měla nahradit metodu callSuper. Metoda volá
		metodu předka stejného názvu. Pro tuto funkčnost využívá vlastnost
		arguments.caller, která funguje ve všech podporovaných prohlížečích
		(Opera od verze 9.6)
	dom.js - SZN.Dom (verze: 3.4)
	  - upravena metoda getBoxPosition() - pokud není zadán druhý parametr,
		je vzat element body
	  - metoda elementHider() přepsána tak, aby si schované elementy uschovala
		do pole, které je vlastnosti funkce a ne aby si označovala skryté elementy
		pomocí vlastního atributu.
	  - metoda getElementsByClass() používá DOM metodu getElementsByClassName
		tam kde tato metoda existuje
	  - přidána metoda arrayFromCollection(), která z živé kolekce, kterou vrací
		např. document.geElementsByTagName vytvoří pole
	events.js - SZN.Events (verze: 2.1)

browser.js
/**
* @overview detekce prohlizece
* @version 1.0
* @author : jelc
*/   


/**
* @class Detekce klientskeho prostredi v zavislosti na vlastnostech javascriptu
* (pokud je to mozne jinak dle vlastnosti navigator.userAgent)
*
*/
SZN.Browser = function(){
	this.Browser();
};

SZN.Browser.Name = 'Browser';
SZN.Browser.version = '1.0';

/** @field {string} platform system uzivatele */
SZN.Browser.prototype.platform = '';
/** @field {string} klient prohlizec uzivatele */
SZN.Browser.prototype.klient = '';
/** @field {string} version verze prohlizece */
SZN.Browser.prototype.version = 0;
/** @field {string} agent hodnota systemove promene "navigator.userAgent" */
SZN.Browser.prototype.agent = '';

/**
* @method implicitni konkstruktor
*/   
SZN.Browser.prototype.Browser = function(){
	this._agent = navigator.userAgent;
	this._platform = navigator.platform;
	this._vendor = navigator.vendor;
	this.platform = this._getPlatform();
	this.klient = this._getKlient();
	this.version = this._getVersion();
	this.mouse = this._getMouse();
};

/**
* @private
* @method zjistuje system uzivatele
* @returns {string} ktery popisuje pouzivany system:
* <ul>
* <li>nix - Linux, BSD apod.</li>
* <li>mac - Apple</li>
* <li>win - Windows pro PC</li>
* <li>oth - vsechno ostatni</li>  
* </ul>    
*
*/   
SZN.Browser.prototype._getPlatform = function(){
	if((this._agent.indexOf('X11') != -1) 
	|| (this._agent.indexOf('Linux') != -1)){
		return 'nix';
	} else if(this._agent.indexOf('Mac') != -1){
		return 'mac';
	} else if(this._agent.indexOf('Win') != -1){
		return 'win';
	} else {
		return 'oth';
	}
};

/**
* @private
* @method zjistuje typ prohlizece
* @returns {string} ktery popisuje pouzivany prohlizec
* <ul>
* <li>opera - Opera</li>
* <li>ie - Internet Explorer</li>
* <li>gecko - Mozilla like</li>
* <li>konqueror - Konqueror</li>  
* <li>safari - Safari</li>  
* <li>oth - vsechno ostatni/neznamy</li>  
* </ul>  
*/   
SZN.Browser.prototype._getKlient = function(){
	if(window.opera){
		return 'opera';
	} else if(document.attachEvent 
	&& (typeof navigator.systemLanguage != 'undefined')){
		return 'ie';
	} else if (document.getAnonymousElementByAttribute){
		return 'gecko';
	} else if(this._agent.indexOf('KHTML')){
		if(this._vendor == 'KDE'){
			return 'konqueror';
		} else {
			return 'safari';
		}
	} else {
		return 'oth';
	}
};

/**
* @private 
* @method Nastavuje identifikaci leveho a praveho tlacitka mysi
* @returns {object} jako asociativni pole s vlastnostmi
* <em>left</em> a <em>right</em>, ktere obsahuji ciselny identifikator
* stisknuteho tlacitka mysi jak ho klient vraci ve vlastnosti udalosti
* <em>e.button</em>
*/

SZN.Browser.prototype._getMouse = function(){
	var left;
	var right;
	if ((this.klient == 'ie') || (this.klient == 'konqueror') || (this.klient == 'safari')){
		left = 1;
		right = 2;
	} else if((this.klient == 'opera') && (this.version > 7) && (this.version < 9)) {
		left = 1;
		right = 2;		
	} else {
		left = 0;
		right = 2;
	}
	return {left:left,right:right};	
}

/**
* @private
* @method zjistuje verzi daneho prohlizece, detekovaneho metodou "_getKlient"
* @returns {string} navratova hodnota metod jejich nazev je slozeny z retezcu
* '_get_' + vlastnost <em>klient</em>  + '_ver'
* @example  <pre>
* pro Internet Exlporer je volana metoda <em>this._get_ie_ver()</em>
*</pre>    
*/   
SZN.Browser.prototype._getVersion = function(){
	var out = 0;
	var fncName = '_get_' + this.klient + '_ver';
	
	if(typeof this[fncName] == 'function'){
		return this[fncName]();
	} else {
		return 0;
	}
};

/**
* @private
* @method detekce verze Internet Exploreru
* @returns {string} verze prohlizece od 5.0 do 7 (IE 8 bude detekovano jako 7)
*/   
SZN.Browser.prototype._get_ie_ver = function(){
	if(typeof Function.prototype.call != 'undefined'){
		if(window.XMLHttpRequest){
			return '7';
		} else if (typeof document.doctype == 'object'){
			return '6';
		} else {
			return '5.5';
		}
	} else {
		return '5.0';
	}
};

/**
* @private
* @method detekce verze Opery
* @returns {string} verze prohlizece od 6 do 9 (> 9 bude detekovano jako 9)
*/  
SZN.Browser.prototype._get_opera_ver = function(){
	if(document.designMode && document.execCommand){
		return '9';
	} else if((document.selection) && (document.createRange)){
		return '8';
	} else if(document.createComment){
		return '7';
	} else {
		return '6';
	}
};

/**
* @private
* @method detekce verze Gecko prohlizecu
* @returns {string} verze prohlizece od 1.5do 2 (> 2 bude detekovano jako 2)
*/ 
SZN.Browser.prototype._get_gecko_ver = function(){
	if(window.external){
		return '2';
	} else {
		return '1.5';
	}
};

/**
* @private
* @method detekce verze Konqueroru
* @returns {string} verze prohlizece na zaklade hodnot uvedenych v navigator.userAgent
* detekuji se prvni dve cisla (3.4,3.5,3.6 atd...) 
*/ 
SZN.Browser.prototype._get_konqueror_ver = function(){
	var num = this._agent.indexOf('KHTML') + 6;
	var part =  this._agent.substring(num);
	var end = part.indexOf(' ')
	var x = part.substring(0,end - 2);
	return x;
	
};
/**
 * @private
 * @method detekce verze Safari
 * @returns {string} verze verze se nedetekuje vraci 1
 */ 
SZN.Browser.prototype._get_safari_ver = function(){
	return '1';
};
